prqlc/prql-compiler/src/sql/srq/preprocess.rs

```diff
@@ -173,8 +173,8 @@ pub(in crate::sql) fn distinct(
                     // DISTINCT
 
                     res.push(SqlTransform::Distinct);
-                } else if ctx.dialect.supports_distinct_on() {
-                    // DISTINCT ON
+                } else if ctx.dialect.supports_distinct_on() && range_int.end.unwrap_or(1) == 1 {
```

Comment:
This will not work correctly if you `take n..` (no upper bound in a range).

I've pushed a test that fails in this case.

Reply:
Ah excellent catch! Thanks for the test I'll have a look into it